File: backup/Twitch/app/src/main/java/com/exact/twitch/ui/player/video/VideoPlayerFragment.java
package com.exact.twitch.ui.player.video;

import androidx.lifecycle.ViewModelProviders;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;

import com.exact.twitch.R;
import com.exact.twitch.ui.VideoDownloadDialog;
import com.exact.twitch.ui.fragment.RadioButtonBottomSheetDialogFragment;
import com.exact.twitch.ui.player.BasePlayerFragment;
import com.exact.twitch.ui.player.PlayerViewModel;
import com.exact.twitch.util.FragmentUtils;
import com.google.android.exoplayer2.source.hls.playlist.RenditionKey;

import java.util.LinkedList;
import java.util.List;

public class VideoPlayerFragment extends BasePlayerFragment implements RadioButtonBottomSheetDialogFragment.OnOptionSelectedListener, VideoDownloadDialog.OnDownloadClickListener {

    private VideoPlayerViewModel viewModel;
//    private ChatRecyclerView chatRecyclerView;
private ImageButton downloadBtn;
    private ImageButton settingsBtn;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_player_video, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
//        chatRecyclerView = view.findViewById(R.id.fragment_player_chat_rv);
        downloadBtn = view.findViewById(R.id.player_download);
        settingsBtn = view.findViewById(R.id.player_settings);
//        channelBtn.setOnClickListener(v -> channelListener.viewChannel(video.getChannel().getName()));
        settingsBtn.setOnClickListener(v -> {
            LinkedList<CharSequence> list = new LinkedList<>(viewModel.getQualities().getValue());
            list.addFirst(getString(R.string.auto));
            FragmentUtils.INSTANCE.showRadioButtonDialogFragment(
                    requireActivity(),
                    getChildFragmentManager(),
                    list,
                    viewModel.getSelectedQualityIndex());
        });
        if (downloadBtn != null)
        downloadBtn.setOnClickListener(v -> new VideoDownloadDialog(this, viewModel.getQualities().getValue(), viewModel.getSegments()).show());
        //TODO morebtn
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        viewModel = ViewModelProviders.of(this, viewModelFactory).get(VideoPlayerViewModel.class);
        pvPlayer.setPlayer(viewModel.getPlayer());
        if (viewModel.isFirstLaunch()) {
            settingsBtn.setEnabled(false);
            if (downloadBtn != null)
            downloadBtn.setEnabled(false);
            viewModel.setVideo(getArguments().getParcelable("video"));
            viewModel.getQualities().observe(this, list -> {
                boolean enabled = list != null;
                if (downloadBtn != null)
                downloadBtn.setEnabled(enabled);
                settingsBtn.setEnabled(enabled);
            });
            viewModel.play();
        }
//        viewModel.getChatMessages().observe(this, chatRecyclerView::submitList);
//        viewModel.getNewMessage().observe(this, message -> chatRecyclerView.notifyAdapter());
    }

    @Override
    public void onStop() {
        super.onStop();
        if (!requireActivity().isChangingConfigurations()) {
            viewModel.getPlayer().release();
        }
    }

    @Override
    public void onSelect(int optionId, CharSequence optionText) {
        if (viewModel.getSelectedQualityIndex() != optionId) {
            if (optionId == 0) {
                viewModel.enableAutoMode();
            } else if (optionId < viewModel.getQualities().getValue().size()) {
                viewModel.changeQuality(optionId - 1);
                viewModel.setSelectedQualityIndex(optionId);
            } else {
                viewModel.setPlayerMode(PlayerViewModel.PlayerMode.AUDIO_ONLY);
            }
        }
    }

    @Override
    public void onClick(String quality, List<RenditionKey> keys) {
        viewModel.download(quality, keys);
    }
}

File: app/src/main/java/com/exact/xtra/ui/player/video/VideoPlayerFragment.kt
package com.exact.xtra.ui.player.video

import android.os.Bundle
import android.os.Parcelable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.exact.xtra.R
import com.exact.xtra.ui.VideoDownloadDialog
import com.exact.xtra.ui.fragment.RadioButtonDialogFragment
import com.exact.xtra.ui.player.BasePlayerFragment
import com.exact.xtra.util.FragmentUtils
import kotlinx.android.synthetic.main.fragment_player_video.*
import kotlinx.android.synthetic.main.player_video.*
import java.util.*

class VideoPlayerFragment : BasePlayerFragment(), RadioButtonDialogFragment.OnSortOptionChanged, VideoDownloadDialog.OnDownloadClickListener {
    override fun play(obj: Parcelable) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    private companion object {
        const val TAG = "VideoPlayer"
    }

    override lateinit var viewModel: VideoPlayerViewModel

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_player_video, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        //        channelBtn.setOnClickListener(v -> channelListener.viewChannel(video.getChannel().getName()));
        //TODO morebtn
        settings.isEnabled = false
        download.isEnabled = false
        settings.setOnClickListener {
            println(viewModel.videoInfo.targetDuration)
            LinkedList(viewModel.helper.qualities.value).also { list ->
                list.addFirst(getString(R.string.auto))
                FragmentUtils.showRadioButtonDialogFragment(childFragmentManager, list, viewModel.helper.selectedQualityIndex)
            }
        }
        download.setOnClickListener { VideoDownloadDialog.newInstance(viewModel.videoInfo).show(childFragmentManager, null) }
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        viewModel = ViewModelProviders.of(this, viewModelFactory).get(VideoPlayerViewModel::class.java)
        playerView.player = viewModel.player
        viewModel.helper.qualities.observe(this, Observer {
            val loaded = it != null
            settings.isEnabled = loaded
            download.isEnabled = loaded
        })
//        viewModel.helper.chatMessages.observe(this, Observer(chatView::submitList))
//        viewModel.helper.newMessage.observe(this, Observer { chatView.notifyAdapter() })
        if (!viewModel.isInitialized()) {
            viewModel.video = arguments!!.getParcelable("video")!!
            viewModel.init()
        }
    }

    override fun onChange(index: Int, text: CharSequence, tag: Int?) {
        viewModel.changeQuality(index)
    }

    override fun onClick(quality: String, segmentFrom: Int, segmentTo: Int) {
        viewModel.download(quality, segmentFrom, segmentTo)
    }
}
